src/com/sun/jna/Memory.java

```diff
@@ -40,91 +44,15 @@
  *        free(buf);
  * </pre>
  *
- * <p>The {@link #finalize} method will free allocated memory when
- * this object is no longer referenced.
- *
  * @author Sheng Liang, originator
  * @author Todd Fast, suitability modifications
  * @author Timothy Wall
  * @see Pointer
  */
-public class Memory extends Pointer {
-
-    private static ReferenceQueue<Memory> QUEUE = new ReferenceQueue<Memory>();
-    private static LinkedReference HEAD; // the head of the doubly linked list used for instance tracking
-
-    /**
-     * Keep track of all allocated memory so we can dispose of it before
-     * unloading. This is done using a doubly linked list to enable fast
-     * removal of tracked instances.
-     */
-    private static class LinkedReference extends WeakReference<Memory> {
-
-        private LinkedReference next;
-        private LinkedReference prev;
-
-        private LinkedReference(Memory referent) {
-            super(referent, QUEUE);
-        }
-
-        /**
-         * Add the given {@code instance} to the instance tracking.
-         *
-         * @param instance the instance to track
-         */
-        static LinkedReference track(Memory instance) {
-            // use a different lock here to allow the finialzier to unlink elements too
-            synchronized (QUEUE) {
-                LinkedReference stale;
-
-                // handle stale references here to avoid GC overheating when memory is limited
-                while ((stale = (LinkedReference) QUEUE.poll()) != null) {
-                    stale.unlink();
-                }
-            }
-
-            // keep object allocation outside the syncronized block
-            LinkedReference entry = new LinkedReference(instance);
-
-            synchronized (LinkedReference.class) {
-                if (HEAD != null) {
-                    entry.next = HEAD;
-                    HEAD = HEAD.prev = entry;
-                } else {
-                    HEAD = entry;
-                }
-            }
-
-            return entry;
-        }
-
-        /**
-         * Remove the related instance from tracking and update the linked list.
-         */
-        private void unlink() {
-            synchronized (LinkedReference.class) {
-                LinkedReference next;
-
-                if (HEAD != this) {
-                    if (this.prev == null) {
-                        // this entry was detached before, e.g. disposeAll was called and finalizers are running now
-                        return;
-                    }
-
-                    next = this.prev.next = this.next;
-                } else {
-                    next = HEAD = HEAD.next;
-                }
-
-                if (next != null) {
-                    next.prev = this.prev;
-                }
-
-                // set prev to null to detect detached entries
-                this.prev = null;
-            }
-        }
-    }
+public class Memory extends Pointer implements Closeable {
```

Comment:
`AutoCloseable` would really be preferred but that would require JDK7.  

I'm wondering if it would be good to also make some of the key JNA classes that hold an internal `Memory` object `Closeable` as well, having their `close()` method invoke the memory's `close()`.  I'm thinking of `Structure` and `ByReference` in particular.

I don't think it would break compatibility but will let downstream projects use Structure and FooByReference allocations in try-with-resources blocks.

Reply:
I'm inclined to say yes, but IMHO this is out of scope for this change.

Reply:
Agreed. Once this is merged I might put together a new PR to add this.